tracing/ops_tracing/buffer.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+"""Buffer for tracing data in the ops-tracing extension."""
+
+from __future__ import annotations
+
+import contextlib
+import functools
+import logging
+import sqlite3
+from pathlib import Path
+from typing import Callable
+
+from typing_extensions import ParamSpec, TypeVar
+
+from .const import Config
+
+# Approximate safety limit for the database file size
+BUFFER_SIZE = 40 * 1024 * 1024
+
+# Default priority for tracing data.
+# Dispatch invocation that doesn't result in any event being observed
+# by charm or its charm lib produces data at this priority.
+DEFAULT_PRIORITY = 10
+
+# Higher priority for data from invocation with observed events.
+OBSERVED_PRIORITY = 50
+
+# Must have a short timeout when terminating
+# May want to have a longer timeout otherwise
+DB_TIMEOUT = 5
+LONG_DB_TIMEOUT = 3600
+
+# Must use isolation_level=None for consistency between Python 3.8 and 3.12
+# Can't use the STRICT keyword for tables, requires sqlite 3.37.0
+# Can't use the octet_length() either, requires sqlite 3.43.0
+# Can't use DELETE ... RETURNING, requires sqlite 3.35.0
+#
+# Ubuntu 20.04  Python  3.8.2  Sqlite 3.31.1  Adds UPSERT, window functions
+# Ubuntu 22.04  Python 3.10.x  Sqlite 3.37.2  Adds STRICT tables, JSON ops
+# Ubuntu 24.04  Python 3.12.x  Sqlite 3.45.2  Adds math functions
+
+logger = logging.getLogger(__name__)
+
+P = ParamSpec('P')
+R = TypeVar('R')
+
+
+def retry(f: Callable[P, R]) -> Callable[P, R]:
+    """Simple retry decorator."""
+
+    @functools.wraps(f)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+        exc: sqlite3.Error | None = None
+
+        for _ in range(3):
+            try:
+                return f(*args, **kwargs)
+            except sqlite3.Error as e:  # noqa: PERF203
+                exc = e
+                continue
+        else:
+            assert exc
+            raise exc
+
+    return wrapper
+
+
+class Buffer:
+    """Buffer for tracing data.
+
+    Access buffer attributes is effectively protected by an sqlite transaction.
+    """
+
+    ids: set[int]
+    """tracing data ids buffered during this dispatch invocation."""
```

Comment:
Maybe

```suggestion
    """Tracing data ids buffered during this dispatch invocation."""
```

Unless it's "tracing" because that's the name of the package? If so then probably tracing in backticks?